chain-signatures/node/src/storage/presignature_storage.rs

```diff
@@ -49,43 +103,173 @@ impl PresignatureStorage {
+            local our_shares = redis.call("SMEMBERS", owner_key)
+            for _, id in ipairs(our_shares) do
+                if not owner_shares[id] then
+                    table.insert(outdated, id)
+                end
+            end
+
+            -- remove the outdated shares from our node
+            if #outdated > 0 then
+                redis.call("SREM", owner_key, unpack(outdated))
+                redis.call("SREM", reserved_key, unpack(outdated))
+                redis.call("HDEL", presig_key, unpack(outdated))
+            end
+
+            return outdated
+        "#;
+
+        let result: Result<Vec<PresignatureId>, _> = redis::Script::new(script)
+            .key(&self.presig_key)
+            .key(&self.reserved_key)
+            .key(owner_key(&self.owner_keys, owner))
+            // NOTE: this encodes each entry of owner_shares as a separate ARGV[index] entry.
+            .arg(owner_shares)
+            .invoke_async(&mut conn)
+            .await;
+
+        match result {
+            Ok(outdated) => {
+                if !outdated.is_empty() {
+                    tracing::info!(?outdated, "removed outdated presignatures");
+                }
+                outdated
+            }
+            Err(err) => {
+                tracing::warn!(?err, "failed to remove outdated presignatures");
+                Vec::new()
+            }
+        }
     }
 
     /// Insert a presignature into the storage. If `mine` is true, the presignature will be
     /// owned by the current node. If `back` is true, the presignature will be marked as unused.
     pub async fn insert(
         &self,
         presignature: Presignature,
+        owner: Participant,
         mine: bool,
-        back: bool,
     ) -> StoreResult<()> {
         let mut conn = self.connect().await?;
 
         let script = r#"
             local mine_key = KEYS[1]
             local presig_key = KEYS[2]
             local used_key = KEYS[3]
+            local reserved_key = KEYS[4]
+            local owner_keys = KEYS[5]
+            local owner_key = KEYS[6]
             local presig_id = ARGV[1]
             local presig_value = ARGV[2]
             local mine = ARGV[3]
-            local back = ARGV[4]
 
-            if back == "true" then
-                redis.call("HDEL", used_key, presig_id)
-            elseif redis.call('HEXISTS', used_key, presig_id) == 1 then
-                return {err = 'Presignature ' .. presig_id .. ' is already used'}
+            -- if the presignature has NOT been reserved, then something went wrong when acquiring the
+            -- reservation for it via presignature slot.
+            if redis.call("SREM", reserved_key, presig_id) == 0 then
+                return {err = "WARN presignature " .. presig_id .. " has NOT been reserved"}
+            end
+
+            if redis.call('HEXISTS', used_key, presig_id) == 1 then
+                return {err = 'WARN presignature ' .. presig_id .. ' is already used'}
```

Comment:
"... already inserted", "used" can be misunderstood

Reply:
not sure what you mean. Used and inserted are two distinct states of stored items, so I'm not sure how they can be misunderstood, plus we don't log whether something is inserted but `stored`.